File: Data Structures & Algorithms/permutations/submission-0.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        ans = []
        path = []

        def dfs(i, s):
            if i == n:
                ans.append(path.copy())
            
            for num in s:
                path.append(num)
                dfs(i + 1, s - {num})
                path.pop()
        
        dfs(0, set(nums))
        return ans

File: Data Structures & Algorithms/combination-target-sum/submission-0.cpp
class Solution {
public:
    void backtrack(
        int start, 
        vector<int>& nums, 
        int target,
        vector<int>& path,
        vector<vector<int>>& ans
    ) {
        if (target < 0) {
            return;
        }

        if (target == 0) {
            ans.push_back(path);
            return;
        }

        for (int i = start; i < nums.size(); i++) {
            path.push_back(nums[i]);
            backtrack(i, nums, target - nums[i], path, ans);
            path.pop_back();
        }
    }

    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<int> path;
        vector<vector<int>> ans;
        backtrack(0, nums, target, path, ans);

        return ans;
    }
};
